### PR TITLE
Add initial_query_start_time to query log

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now query_log has two new columns : initial_query_start_time / initial_query_start_time_microsecond that record the starting time of a distributed query if any.

resubmit https://github.com/ClickHouse/ClickHouse/pull/24388

This data race shows everytime when we try to extend query_log. The fix is to guard field initialization via the `internal` flag.

Detailed description / Documentation draft:

.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
What's wrong with the `01103_optimize_drop_race_zookeeper` test...

**Comment by amosbird** [comment]
> What's wrong with the `01103_optimize_drop_race_zookeeper` test...

I don't know... From the test content it's not related to this PR.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
@Mergifyio update

File: src/Interpreters/ClientInfo.cpp [hunk 1]
```diff
@@ -34,6 +34,12 @@ void ClientInfo::write(WriteBuffer & out, const UInt64 server_protocol_revision)
     writeBinary(initial_query_id, out);
     writeBinary(initial_address.toString(), out);
 
+    if (server_protocol_revision >= DBMS_MIN_PROTOCOL_VERSION_WITH_INITIAL_QUERY_START_TIME)
+    {
+        writeBinary(initial_query_start_time, out);
```
Comment (alexey-milovidov) [inline src/Interpreters/ClientInfo.cpp:39]: Let's not serialize/deserialize both time and time with microseconds as this is redundant. We can serialize just the initial_query_start_time_microseconds and reconstruct the other.

File: src/Interpreters/executeQuery.cpp [hunk 2]
```diff
@@ -355,6 +355,15 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
 {
     const auto current_time = std::chrono::system_clock::now();
 
+    auto & client_info = context->getClientInfo();
+
+    // If it's an initial query, set to current_time
```
Comment (alexey-milovidov) [inline src/Interpreters/executeQuery.cpp:360]: The reader will wonder - why it's only for non-internal queries.

PS. `If it's an initial query` -> `If it's not an initial query`.

File: src/Interpreters/executeQuery.cpp [hunk 3]
```diff
@@ -355,6 +355,15 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
...
+    if (!internal && client_info.initial_query_start_time == 0)
```
Comment (alexey-milovidov) [inline src/Interpreters/executeQuery.cpp:361]: Checking `QueryKind::INITIAL_QUERY` might be more natural.

File: src/Interpreters/executeQuery.cpp (hunk 2 above)
Comment (amosbird) [reply src/Interpreters/executeQuery.cpp:360]: The comment indeed needs modification: `If it's not an internal query, and the initial query time is not yet set`

Internal queries are those executed without a client context, which should not have initial_query_start_time set.

It's possible to not have initial query time set for non-internal and non-initial queries. For example, the initiator is running an old version of clickhouse that don't have the feature yet.

File: src/Interpreters/executeQuery.cpp (hunk 3 above)
Comment (amosbird) [reply src/Interpreters/executeQuery.cpp:361]: I'll add proper comments to explain it.

File: src/Interpreters/ClientInfo.cpp (hunk 1 above)
Comment (amosbird) [reply src/Interpreters/ClientInfo.cpp:39]: Done

**Review by alexey-milovidov — APPROVED** [review]

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-10297.

**Pull request merged**